modules/core/src/main/java/org/apache/ignite/internal/cdc/ChangeDataCapture.java

```diff
@@ -0,0 +1,531 @@
+
+        marshaller = MarshallerContextImpl.mappingFileStoreWorkDir(cfg.getWorkDirectory());
+
+        if (log.isDebugEnabled()) {
+            log.debug("Using BinaryMeta directory[dir=" + binaryMeta + ']');
+            log.debug("Using Marshaller directory[dir=" + marshaller + ']');
+        }
+
+        injectResources(consumer.getCdcConsumer());
+    }
+
+    /** Waits and consumes new WAL segments until stoped. */
+    public void consumeWalSegmentsUntilStopped() {
+        try {
+            Set<Path> seen = new HashSet<>();
+
+            long[] lastSgmnt = new long[] {-1};
+
+            while (!stopped) {
+                try (Stream<Path> cdcFiles = Files.walk(cdcDir, 1)) {
+                    Set<Path> exists = new HashSet<>();
+
+                    cdcFiles
+                        .peek(exists::add) // Store files that exists in cdc dir.
+                        .filter(WAL_SEGMENTS_FILTER.and(p -> !seen.contains(p))) // Need unseend WAL segments only.
+                        .peek(seen::add) // Adds to seen.
+                        .sorted(Comparator.comparingLong(this::segmentIndex)) // Sort by segment index.
+                        .peek(p -> {
+                            long nextSgmnt = segmentIndex(p);
+
+                            assert lastSgmnt[0] == -1 || nextSgmnt - lastSgmnt[0] == 1;
+
+                            lastSgmnt[0] = nextSgmnt;
+                        })
+                        .forEach(this::consumeSegment); // Consuming segments.
+
+                    seen.removeIf(p -> !exists.contains(p)); // Clean up seen set.
+                }
+
+                if (!stopped)
+                    U.sleep(cdcCfg.getCheckFrequency());
+            }
+        }
+        catch (IOException | IgniteInterruptedCheckedException e) {
+            throw new IgniteException(e);
+        }
+    }
+
+    /** Reads all available records from segment. */
+    private void consumeSegment(Path segment) {
+        if (log.isInfoEnabled())
+            log.info("Processing WAL segment[segment=" + segment + ']');
+
+        IgniteWalIteratorFactory.IteratorParametersBuilder builder = new IgniteWalIteratorFactory.IteratorParametersBuilder()
+            .log(log)
+            .binaryMetadataFileStoreDir(binaryMeta)
+            .marshallerMappingFileStoreDir(marshaller)
+            .keepBinary(cdcCfg.isKeepBinary())
+            .filesOrDirs(segment.toFile())
+            .addFilter((type, ptr) -> type == DATA_RECORD_V2);
+
+        if (initState != null) {
+            long segmentIdx = segmentIndex(segment);
+
+            if (segmentIdx > initState.index()) {
+                log.error("Found segment greater then saved state. Some events are missed. Exiting!" +
+                    "[state=" + initState + ",segment=" + segmentIdx + ']');
+
+                throw new IgniteException("Some data missed.");
+            }
+
+            if (segmentIdx < initState.index()) {
```

Comment:
Seem to be a critical error too.
The consumer already deleted this file.

Reply:
It May be not an error, but we should have a single remove point, I think.

Reply:
No, if app failed after state saved but before processed segment deleted.

```
                if (commit) {
                    assert it.lastRead().isPresent();

                    state.save(it.lastRead().get());

// CRUSH HERE!

                    if (!prevSegments.isEmpty()) {
                        for (Path prevSegment : prevSegments)
                            Files.deleteIfExists(prevSegment);

                        prevSegments.clear();
                    }
                }

```